### PR TITLE
protoc-gen-openapiv2: Use json_name when generating required field names

### LINKED ISSUE #2874: protoc-gen-openapiv2 does not respect json_name when generating the required properties
## 🐛 Bug Report
Starting with protoc-gen-openapiv2 v2.8.0, the generated OpenAPIv2 schema is generated incorrectly when using the `grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema` option, together with `json_name`. The properties section uses the name from the json_name annotation, but the required section uses the autogenerated name instead.  

## To Reproduce

I've tested this with multiple versions, the behavior between latest (v2.11.3 currently) and v2.8.0 does not change, this is the first version with the current behavior.

Protobuf file:

```protobuf
syntax = "proto3";

option go_package = "pkg/api";

import "protoc-gen-openapiv2/options/annotations.proto";
import "google/api/annotations.proto";

service Test {
  rpc Test(Required) returns (Required) {
      option (google.api.http) = {
        get: "/test"
      };
  };
}

message Required {
  option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_schema) = {
    json_schema: {
      required: [ "test_id"]
    }
  };
  string test_id = 1 [json_name="test_id"];
}
```

Project setup for v2.8.0:
```sh
go install github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway@v2.8.0 github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-openapiv2@v2.8.0
mkdir -p ./protoc-gen-openapiv2/options/
curl -o ./protoc-gen-openapiv2/options/annotations.proto https://raw.githubusercontent.com/grpc-ecosystem/grpc-gateway/v2.8.0/protoc-gen-openapiv2/options/annotations.proto
curl -o ./protoc-gen-openapiv2/options/openapiv2.proto https://raw.githubusercontent.com/grpc-ecosystem/grpc-gateway/v2.8.0/protoc-gen-openapiv2/options/openapiv2.proto
mkdir -p ./google/api/
curl -o ./google/api/http.proto https://raw.githubusercontent.com/googleapis/googleapis/master/google/api/http.proto
curl -o ./google/api/annotations.proto https://raw.githubusercontent.com/googleapis/googleapis/master/google/api/annotations.proto
protoc -I . --openapiv2_out ./ required.proto
```

Project setup for v2.7.3:
```sh
go install github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-grpc-gateway@v2.7.3 github.com/grpc-ecosystem/grpc-gateway/v2/protoc-gen-openapiv2@v2.7.3
mkdir -p ./protoc-gen-openapiv2/options/
curl -o ./protoc-gen-openapiv2/options/annotations.proto https://raw.githubusercontent.com/grpc-ecosystem/grpc-gateway/v2.7.3/protoc-gen-openapiv2/options/annotations.proto
curl -o ./protoc-gen-openapiv2/options/openapiv2.proto https://raw.githubusercontent.com/grpc-ecosystem/grpc-gateway/v2.7.3/protoc-gen-openapiv2/options/openapiv2.proto
mkdir -p ./google/api/
curl -o ./google/api/http.proto https://raw.githubusercontent.com/googleapis/googleapis/master/google/api/http.proto
curl -o ./google/api/annotations.proto https://raw.githubusercontent.com/googleapis/googleapis/master/google/api/annotations.proto
protoc -I . --openapiv2_out ./ required.proto
```

## Expected behavior

The generated OpenAPIv2 spec is generated like in v2.7.3:

```json
{
  "swagger": "2.0",
  "info": {
    "title": "required.proto",
    "version": "version not set"
  },
  "tags": [
    {
      "name": "Test"
    }
  ],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "paths": {
    "/test": {
      "get": {
        "operationId": "Test_Test",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/Required"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "test_id",
            "in": "query",
            "required": false,
            "type": "string"
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },
  "definitions": {
    "Required": {
      "type": "object",
      "properties": {
        "test_id": {
          "type": "string"
        }
      },
      "required": [
        "test_id"
      ]
    },
    "protobufAny": {
      "type": "object",
      "properties": {
        "@type": {
          "type": "string"
        }
      },
      "additionalProperties": {}
    },
    "rpcStatus": {
      "type": "object",
      "properties": {
        "code": {
          "type": "integer",
          "format": "int32"
        },
        "message": {
          "type": "string"
        },
        "details": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/protobufAny"
          }
        }
      }
    }
  }
}

```

## Actual Behavior

Instead the required section of the generated schema references a property with the wrong (autogenerated) name:

```json
{
  "swagger": "2.0",
  "info": {
    "title": "required.proto",
    "version": "version not set"
  },
  "tags": [
    {
      "name": "Test"
    }
  ],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "paths": {
    "/test": {
      "get": {
        "operationId": "Test_Test",
        "responses": {
          "200": {
            "description": "A successful response.",
            "schema": {
              "$ref": "#/definitions/Required"
            }
          },
          "default": {
            "description": "An unexpected error response.",
            "schema": {
              "$ref": "#/definitions/rpcStatus"
            }
          }
        },
        "parameters": [
          {
            "name": "test_id",
            "in": "query",
            "required": false,
            "type": "string"
          }
        ],
        "tags": [
          "Test"
        ]
      }
    }
  },
  "definitions": {
    "Required": {
      "type": "object",
      "properties": {
        "test_id": {
          "type": "string"
        }
      },
      "required": [
        "testId"
      ]
    },
    "protobufAny": {
      "type": "object",
      "properties": {
        "@type": {
          "type": "string"
        }
      },
      "additionalProperties": {}
    },
    "rpcStatus": {
      "type": "object",
      "properties": {
        "code": {
          "type": "integer",
          "format": "int32"
        },
        "message": {
          "type": "string"
        },
        "details": {
          "type": "array",
          "items": {
            "$ref": "#/definitions/protobufAny"
          }
        }
      }
    }
  }
}
```

Diff between expected and actual:

```diff
--- expected.json       2022-09-06 13:02:03.000000000 +0200
+++ actual.json 2022-09-06 12:56:06.000000000 +0200
@@ -56,7 +56,7 @@
         }
       },
       "required": [
-        "test_id"
+        "testId"
       ]
     },
     "protobufAny": {
```

## Your Environment
MacOS Big Sur (11.6.8), go1.19 darwin/amd64, grpc-gateway versions v2.11.3, v2.8.0 and v2.7.3

### PR DESCRIPTION
**Pull request by patrick246**
#### References to other Issues or PRs

Fixes #2874

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md)?
yes

#### Brief description of what is fixed or changed

Fix the bug from #2874 where the required fields section in the JSON schema would contain autogenerated field names, even when fields had a custom json_name value set. These would then not match up with the actual fields defined in the schema, resulting in an invalid OpenAPI spec.

This adds checks if we need to convert Protobuf field names into JSON field names in template.go `schemaOfField` and `nestedQueryParams` and removes a check from `updateswaggerObjectFromJSONSchema` that assumed no json_name was set and prevented another conversion in `renderMessageAsDefinition` from working.

The PR also contains additions to the tests for functions that were modified, in addition to examples in `a_bit_of_everything.proto`.

### CODE REVIEW

**Comment by patrick246** [comment]
I'll have to debug the failing CLA check tomorrow, there should be a signed corporate CLA in place

**Review by johanbrandhorst — APPROVED** [review]
LGTM

**Comment by johanbrandhorst** [comment]
Looks like we got a formatting diff in the proto changes, you can fix it with `buf format`.

**Comment by johanbrandhorst** [comment]
@oyvindwe could you please have a look at this as well?

File: protoc-gen-openapiv2/internal/genopenapi/template.go [hunk 1]
```diff
@@ -269,7 +269,7 @@ func nestedQueryParams(message *descriptor.Message, field *descriptor.Field, pre
 		// verify if the field is required in message options
 		if messageSchema, err := extractSchemaOptionFromMessageDescriptor(message.DescriptorProto); err == nil {
 			for _, fieldName := range messageSchema.GetJsonSchema().GetRequired() {
-				if fieldName == reg.FieldName(field) {
+				if fieldName == reg.FieldName(field) || fieldName == field.GetName() {
```
Comment (oyvindwe) [inline protoc-gen-openapiv2/internal/genopenapi/template.go:272]: What is the case where `reg.FieldName(field)` and `field.GetName()` would differ?

Comment (patrick246) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:272]: `reg.FieldName()` returns `field.GetName()` or `field.GetJsonName()`, depending on the flag `json_names_for_fields`. The condition could also be replaced with `fieldName == field.GetName() || fieldName == field.GetJsonName()`

Comment (oyvindwe) [reply protoc-gen-openapiv2/internal/genopenapi/template.go:272]: Thanks for the clarification! I think the latter suggestion is easier to understand when reading the code.

As I understand this, then for e.g. field `foo_bar`, `required` could be set to either `foo_bar` or `fooBar` (or a custom JSON name) with the same result. I don't see any problem with providing this flexibility for `required`.

Comment (patrick246) [reply protoc-gen-openapiv2/internal/genopenapi/template.go]: Replaced with a check against field.GetName() and field.GetJsonName()

Comment (oyvindwe) [reply protoc-gen-openapiv2/internal/genopenapi/template.go]: Looks good to me!

**Review by oyvindwe — APPROVED** [review]

File: protoc-gen-openapiv2/internal/genopenapi/template.go (hunk 1 above)
Comment (oyvindwe) [reply protoc-gen-openapiv2/internal/genopenapi/template.go]: @patrick246 And thanks for fixing this issue.

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Comment by patrick246** [comment]
Hi, are there already plans for releasing this?

**Comment by johanbrandhorst** [comment]
Sure, we have a few in-flight PRs (#2928, #2904) but if we don't move forward on those soon I can cut another minor.

**Pull request merged**